Implement encoding and decoding functions

thomashoneyman opened the pull request:
This PR introduces the `encode` and `decode` functions for URIs and for `x-www-form-urlencoded`. It solves [this comment](https://github.com/purescript/purescript-globals/issues/22#issuecomment-743699826) wrt updating libraries to PureScript 0.14.

I've relied on the MDN documentation seen here:

- https://web.archive.org/web/20201206001047/https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent
- https://web.archive.org/web/20201204193427/https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent

I haven't yet added tests for these functions, but this PR at least allows us to bikeshed on implementation & naming.

thomashoneyman committed:
Initial implementation of encoders and decoders

thomashoneyman committed:
Add documentation comments

thomashoneyman commented on spago.dhall:
```diff
@@ -1,5 +1,6 @@
 { name = "js-uri"
-, dependencies = [ "console", "effect", "psci-support" ]
+, dependencies =
+  [ "console", "effect", "functions", "maybe", "psci-support", "strings" ]
```
I can drop the `strings` dependency if I apply replacements in JavaScript instead of in PureScript (for the `*FormURLComponent` functions).

garyb replied:
I'm not against this. This library already has some non-trivial FFI bindings (well, they're only just non-trivial :smile:) so doing it all in the FFI is fine to me.

thomashoneyman committed:
Use strict in FFI

thomashoneyman commented:
Once merged, I can make a release, add this to the 0.14 package set, and then:

1. Deprecate the `globals` library
2. Update `affjax`, `uri`, `routing`, and `form-urlencoded` to rely on this library instead of `globals`
3. Create an issue to replace this library, one day, with an implementation in PureScript that doesn't rely on the FFI so that it can be reused by other backends.

thomashoneyman commented on src/JSURI.purs:
````diff
@@ -0,0 +1,90 @@
+-- | unsafeEncode :: String -> String
+-- | unsafeEncode str = unsafePartial $ fromJust $ encodeURIComponent str
+-- | ```
+encodeURIComponent :: String -> Maybe String
````
I also considered `encodeRFC3896` or `encodeURIRFC3896` given that this isn't strictly `encodeURIComponent` from JavaScript, but rather that function augmented to properly conform to the spec.

hdgarrood commented:
I didn’t realise that the longer term ambition is to have this library written in PureScript with no FFI; in this case I think the name js-uri is actually a bit misleading. Or at least, it will be once that happens.

thomashoneyman commented:
I don't think _this_ library should be rewritten in PureScript, just that it would be ideal for some RFC 3896 encoding / decoding to exist which is not written using the FFI.

JordanMartinez commented:
I'm pretty sure this is just a stopgap solution so we can get v0.14.0 out the door.

thomashoneyman commented:
Unless someone disagrees, once I've added some tests and dropped the `strings` dependency I'd like to publish this library and use it to complete the 0.14 transition for all contrib libraries. I plan to do that sometime tomorrow unless someone voices dissent before then. Thanks!

thomashoneyman committed:
Move to FFI-only implementation and drop strings dependency.

thomashoneyman committed:
Fix FFI definitions
